{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "593b4de8",
   "metadata": {},
   "source": [
    "# Perform differential expression analysis between RA patients\n",
    "\n",
    "**Author: Caroline Larkin**\n",
    "\n",
    "**Research Group: Gottschalk Lab, Department of Immunology, University of Pittsburgh**\n",
    "\n",
    "Notebook based on [link](https://liana-py.readthedocs.io/en/latest/notebooks/targeted.html)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "cd8a540c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/ihome/rgottschalk/cil8/envs/rnaseq/lib/python3.13/site-packages/celltypist/classifier.py:11: FutureWarning: `__version__` is deprecated, use `importlib.metadata.version('scanpy')` instead\n",
      "  from scanpy import __version__ as scv\n",
      "ðŸ“‚ Storing models in /ihome/rgottschalk/cil8/.celltypist/data/models\n",
      "â© Skipping [1/60]: Immune_All_Low.pkl (file exists)\n",
      "â© Skipping [2/60]: Immune_All_High.pkl (file exists)\n",
      "â© Skipping [3/60]: Adult_COVID19_PBMC.pkl (file exists)\n",
      "â© Skipping [4/60]: Adult_CynomolgusMacaque_Hippocampus.pkl (file exists)\n",
      "â© Skipping [5/60]: Adult_Human_MTG.pkl (file exists)\n",
      "â© Skipping [6/60]: Adult_Human_PancreaticIslet.pkl (file exists)\n",
      "â© Skipping [7/60]: Adult_Human_PrefrontalCortex.pkl (file exists)\n",
      "â© Skipping [8/60]: Adult_Human_Skin.pkl (file exists)\n",
      "â© Skipping [9/60]: Adult_Human_Vascular.pkl (file exists)\n",
      "â© Skipping [10/60]: Adult_Mouse_Gut.pkl (file exists)\n",
      "â© Skipping [11/60]: Adult_Mouse_OlfactoryBulb.pkl (file exists)\n",
      "â© Skipping [12/60]: Adult_Pig_Hippocampus.pkl (file exists)\n",
      "â© Skipping [13/60]: Adult_RhesusMacaque_Hippocampus.pkl (file exists)\n",
      "â© Skipping [14/60]: Adult_cHSPCs_Illumina.pkl (file exists)\n",
      "â© Skipping [15/60]: Adult_cHSPCs_Ultima.pkl (file exists)\n",
      "â© Skipping [16/60]: Autopsy_COVID19_Lung.pkl (file exists)\n",
      "â© Skipping [17/60]: COVID19_HumanChallenge_Blood.pkl (file exists)\n",
      "â© Skipping [18/60]: COVID19_Immune_Landscape.pkl (file exists)\n",
      "â© Skipping [19/60]: Cells_Adult_Breast.pkl (file exists)\n",
      "â© Skipping [20/60]: Cells_Fetal_Lung.pkl (file exists)\n",
      "â© Skipping [21/60]: Cells_Human_Tonsil.pkl (file exists)\n",
      "â© Skipping [22/60]: Cells_Intestinal_Tract.pkl (file exists)\n",
      "â© Skipping [23/60]: Cells_Lung_Airway.pkl (file exists)\n",
      "â© Skipping [24/60]: Developing_Human_Brain.pkl (file exists)\n",
      "â© Skipping [25/60]: Developing_Human_Gonads.pkl (file exists)\n",
      "â© Skipping [26/60]: Developing_Human_Hippocampus.pkl (file exists)\n",
      "â© Skipping [27/60]: Developing_Human_Organs.pkl (file exists)\n",
      "â© Skipping [28/60]: Developing_Human_Thymus.pkl (file exists)\n",
      "â© Skipping [29/60]: Developing_Mouse_Brain.pkl (file exists)\n",
      "â© Skipping [30/60]: Developing_Mouse_Hippocampus.pkl (file exists)\n",
      "â© Skipping [31/60]: Fetal_Human_AdrenalGlands.pkl (file exists)\n",
      "â© Skipping [32/60]: Fetal_Human_Pancreas.pkl (file exists)\n",
      "â© Skipping [33/60]: Fetal_Human_Pituitary.pkl (file exists)\n",
      "â© Skipping [34/60]: Fetal_Human_Retina.pkl (file exists)\n",
      "â© Skipping [35/60]: Fetal_Human_Skin.pkl (file exists)\n",
      "â© Skipping [36/60]: Healthy_Adult_Heart.pkl (file exists)\n",
      "â© Skipping [37/60]: Healthy_COVID19_PBMC.pkl (file exists)\n",
      "â© Skipping [38/60]: Healthy_Human_Liver.pkl (file exists)\n",
      "â© Skipping [39/60]: Healthy_Mouse_Liver.pkl (file exists)\n",
      "â© Skipping [40/60]: Human_AdultAged_Hippocampus.pkl (file exists)\n",
      "â© Skipping [41/60]: Human_Colorectal_Cancer.pkl (file exists)\n",
      "â© Skipping [42/60]: Human_Developmental_Retina.pkl (file exists)\n",
      "â© Skipping [43/60]: Human_Embryonic_YolkSac.pkl (file exists)\n",
      "â© Skipping [44/60]: Human_Endometrium_Atlas.pkl (file exists)\n",
      "â© Skipping [45/60]: Human_IPF_Lung.pkl (file exists)\n",
      "â© Skipping [46/60]: Human_Longitudinal_Hippocampus.pkl (file exists)\n",
      "â© Skipping [47/60]: Human_Lung_Atlas.pkl (file exists)\n",
      "â© Skipping [48/60]: Human_PF_Lung.pkl (file exists)\n",
      "â© Skipping [49/60]: Human_Placenta_Decidua.pkl (file exists)\n",
      "â© Skipping [50/60]: Lethal_COVID19_Lung.pkl (file exists)\n",
      "â© Skipping [51/60]: Mouse_Dendritic_Subtypes.pkl (file exists)\n",
      "â© Skipping [52/60]: Mouse_Dentate_Gyrus.pkl (file exists)\n",
      "â© Skipping [53/60]: Mouse_Isocortex_Hippocampus.pkl (file exists)\n",
      "â© Skipping [54/60]: Mouse_Postnatal_DentateGyrus.pkl (file exists)\n",
      "â© Skipping [55/60]: Mouse_Whole_Brain.pkl (file exists)\n",
      "â© Skipping [56/60]: Nuclei_Human_InnerEar.pkl (file exists)\n",
      "â© Skipping [57/60]: Nuclei_Lung_Airway.pkl (file exists)\n",
      "â© Skipping [58/60]: PaediatricAdult_COVID19_Airway.pkl (file exists)\n",
      "â© Skipping [59/60]: PaediatricAdult_COVID19_PBMC.pkl (file exists)\n",
      "â© Skipping [60/60]: Pan_Fetal_Human.pkl (file exists)\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import anndata as ad \n",
    "import scanpy as sc\n",
    "\n",
    "import celltypist\n",
    "from celltypist import models\n",
    "\n",
    "import liana as li\n",
    "\n",
    "# Import DESeq2\n",
    "from pydeseq2.dds import DeseqDataSet\n",
    "from pydeseq2.ds import DeseqStats\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns \n",
    "\n",
    "sample_ids = [\"SA220\", \"SA222\", \"SA227\"]\n",
    "sample_labels = [\"Naive\", \"Resistant\", \"Resistant\"] \n",
    "\n",
    "data_dir = \"./EMATB8322/\"\n",
    "\n",
    "# Download base models from celltypist\n",
    "models.download_models()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3a3daf19",
   "metadata": {},
   "source": [
    "## 1. Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a2dea3eb-fe19-43f2-8048-a91f43a46509",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load celltypist immune model\n",
    "model = models.Model.load(model = \"Immune_All_High.pkl\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "e2d53110",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs Ã— n_vars = 1832 Ã— 16171\n",
       "    obs: 'Group', 'Sample'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ðŸ”¬ Input data has 1832 cells and 16171 genes\n",
      "ðŸ”— Matching reference genes in the model\n",
      "ðŸ§¬ 4688 features used for prediction\n",
      "âš–ï¸ Scaling input data\n",
      "ðŸ–‹ï¸ Predicting labels\n",
      "âœ… Prediction done!\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "View of AnnData object with n_obs Ã— n_vars = 1822 Ã— 16171\n",
       "    obs: 'Group', 'Sample', 'cell_type'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs Ã— n_vars = 5189 Ã— 17673\n",
       "    obs: 'Group', 'Sample'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ðŸ”¬ Input data has 5189 cells and 17673 genes\n",
      "ðŸ”— Matching reference genes in the model\n",
      "ðŸ§¬ 5000 features used for prediction\n",
      "âš–ï¸ Scaling input data\n",
      "ðŸ–‹ï¸ Predicting labels\n",
      "âœ… Prediction done!\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "View of AnnData object with n_obs Ã— n_vars = 5173 Ã— 17673\n",
       "    obs: 'Group', 'Sample', 'cell_type'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs Ã— n_vars = 4774 Ã— 16896\n",
       "    obs: 'Group', 'Sample'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ðŸ”¬ Input data has 4774 cells and 16896 genes\n",
      "ðŸ”— Matching reference genes in the model\n",
      "ðŸ§¬ 4930 features used for prediction\n",
      "âš–ï¸ Scaling input data\n",
      "ðŸ–‹ï¸ Predicting labels\n",
      "âœ… Prediction done!\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "View of AnnData object with n_obs Ã— n_vars = 4757 Ã— 16896\n",
       "    obs: 'Group', 'Sample', 'cell_type'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ads = []\n",
    "for id, label in zip(sample_ids, sample_labels):\n",
    "    cur_ad = sc.read_mtx(filename=f\"{data_dir}{id}/processed_counts.mtx\").transpose()\n",
    "    gene_names = pd.read_csv(f\"{data_dir}{id}/processed_gene_names.csv\", header=None)\n",
    "    cur_ad.var_names = gene_names.values.flatten()\n",
    "    print()#print(gene_names.values.flatten())\n",
    "    barcodes = pd.read_csv(f\"{data_dir}{id}/processed_metadata.csv\", usecols=[\"barcode\"])\n",
    "    cur_ad.obs_names = barcodes.values.flatten()\n",
    "    cur_ad.obs_names_make_unique()\n",
    "    cur_ad.obs[\"Group\"] = [label]* len(barcodes)\n",
    "    cur_ad.obs[\"Sample\"] = [id] * len(barcodes)\n",
    "    display(cur_ad)\n",
    "\n",
    "    # create a temporary copy of adata to preprocess\n",
    "    temp = cur_ad.copy()\n",
    "    # Normalizing to median total counts\n",
    "    sc.pp.normalize_total(temp, target_sum=10000)\n",
    "    # Logarithmize the data\n",
    "    sc.pp.log1p(temp)\n",
    "\n",
    "    predictions = celltypist.annotate(temp, model=model)\n",
    "    new_ad = predictions.to_adata();\n",
    "    cur_ad.obs[\"cell_type\"] = new_ad.obs[\"predicted_labels\"]\n",
    "    cell_counts = cur_ad.obs[\"cell_type\"].value_counts()\n",
    "    # removing cell types with low counts\n",
    "    #print(cell_counts)\n",
    "    keep_cell_types = cell_counts[cell_counts.values > 10]\n",
    "    #print(keep_cell_types.index.values)\n",
    "    cur_ad = cur_ad[cur_ad.obs[\"cell_type\"].isin(keep_cell_types.index.values)]\n",
    "    display(cur_ad)\n",
    "    #break\n",
    "    ads.append(cur_ad); \n",
    "    del temp\n",
    "    del new_ad\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "3b6f55ac",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs Ã— n_vars = 11752 Ã— 15464\n",
       "    obs: 'Group', 'Sample', 'cell_type'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "adata = sc.concat(ads, merge=\"same\", join=\"inner\")\n",
    "adata.obs_names_make_unique()\n",
    "#adata.X.dtype = np.int64\n",
    "adata.X = adata.X.toarray()\n",
    "adata.X[np.isnan(adata.X)] = 0 \n",
    "display(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "f046aa37",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The total number of genes included in these samples is:  15464\n"
     ]
    }
   ],
   "source": [
    "sets = [set(x.var_names) for x in ads]\n",
    "union = list(sets[0].intersection(*sets[1:]))\n",
    "print(\"The total number of genes included in these samples is: \", len(union))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "14d51dc5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "-------------------------------------------------------------------------------------\n",
      "\n",
      "HERE Fibroblasts\n",
      "CELL TYPE:  Fibroblasts ['Naive' 'Resistant']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/scratch/slurm-7854814/ipykernel_385153/3328906579.py:20: DeprecationWarning: ref_level is deprecated and no longer has any effect. It will beremoved in a future release.\n",
      "/scratch/slurm-7854814/ipykernel_385153/3328906579.py:20: DeprecationWarning: design_factors is deprecated and will soon be removed.Please consider providing a formulaic formula using the design argumentinstead.\n",
      "Fitting size factors...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using None as control genes, passed at DeseqDataSet initialization\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... done in 0.57 seconds.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 1.51 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.11 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 2.81 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 1.87 seconds.\n",
      "\n",
      "Calculating cook's distance...\n",
      "... done in 1.41 seconds.\n",
      "\n",
      "Replacing 20 outlier genes.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 0.01 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 0.01 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 0.01 seconds.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 0.82 seconds.\n",
      "\n",
      "Fitting MAP LFCs...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: Group Resistant vs Naive\n",
      "                baseMean  log2FoldChange     lfcSE      stat        pvalue  \\\n",
      "FO538757.2      0.124881        0.709580  0.242665  2.924109  3.454438e-03   \n",
      "AP006222.2      0.184819        0.307082  0.162347  1.891515  5.855569e-02   \n",
      "RP4-669L17.10   0.004709        0.699487  2.025944  0.345265  7.298952e-01   \n",
      "RP11-206L10.9   0.027324        0.687360  0.668524  1.028176  3.038672e-01   \n",
      "FAM87B          0.010583        0.606791  1.052979  0.576261  5.644385e-01   \n",
      "...                  ...             ...       ...       ...           ...   \n",
      "MT-CYB         23.069859       -0.192027  0.020851 -9.209348  3.281106e-20   \n",
      "AL592183.1      0.029974        0.683871  0.671623  1.018237  3.085654e-01   \n",
      "AC007325.4      0.037172        0.548139  0.539205  1.016569  3.093583e-01   \n",
      "AC004556.1      0.078518       -1.554060  0.292649 -5.310324  1.094303e-07   \n",
      "AC240274.1      0.155812        0.336482  0.202221  1.663930  9.612640e-02   \n",
      "\n",
      "                       padj  \n",
      "FO538757.2     9.841943e-03  \n",
      "AP006222.2     1.105356e-01  \n",
      "RP4-669L17.10           NaN  \n",
      "RP11-206L10.9           NaN  \n",
      "FAM87B                  NaN  \n",
      "...                     ...  \n",
      "MT-CYB         2.744622e-19  \n",
      "AL592183.1              NaN  \n",
      "AC007325.4              NaN  \n",
      "AC004556.1     5.013484e-07  \n",
      "AC240274.1     1.588234e-01  \n",
      "\n",
      "[15464 rows x 6 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... done in 2.09 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shrunk log2 fold change & Wald test p-value: Group[T.Resistant]\n",
      "                baseMean  log2FoldChange     lfcSE      stat        pvalue  \\\n",
      "FO538757.2      0.124881        0.566587  0.280376  2.924109  3.454438e-03   \n",
      "AP006222.2      0.184819       -0.129691  0.169100  1.891515  5.855569e-02   \n",
      "RP4-669L17.10   0.004709       -0.001118  0.875506  0.345265  7.298952e-01   \n",
      "RP11-206L10.9   0.027324        0.046824  0.611874  1.028176  3.038672e-01   \n",
      "FAM87B          0.010583       -0.224822  0.633389  0.576261  5.644385e-01   \n",
      "...                  ...             ...       ...       ...           ...   \n",
      "MT-CYB         23.069859       -0.191441  0.020887 -9.209348  3.281106e-20   \n",
      "AL592183.1      0.029974        0.080791  0.630197  1.018237  3.085654e-01   \n",
      "AC007325.4      0.037172       -0.275127  0.490989  1.016569  3.093583e-01   \n",
      "AC004556.1      0.078518       -8.327761  0.580275 -5.310324  1.094303e-07   \n",
      "AC240274.1      0.155812       -0.179954  0.208185  1.663930  9.612640e-02   \n",
      "\n",
      "                       padj  \n",
      "FO538757.2     9.841943e-03  \n",
      "AP006222.2     1.105356e-01  \n",
      "RP4-669L17.10           NaN  \n",
      "RP11-206L10.9           NaN  \n",
      "FAM87B                  NaN  \n",
      "...                     ...  \n",
      "MT-CYB         2.744622e-19  \n",
      "AL592183.1              NaN  \n",
      "AC007325.4              NaN  \n",
      "AC004556.1     5.013484e-07  \n",
      "AC240274.1     1.588234e-01  \n",
      "\n",
      "[15464 rows x 6 columns]\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "-------------------------------------------------------------------------------------\n",
      "\n",
      "HERE Macrophages\n",
      "CELL TYPE:  Macrophages ['Naive' 'Resistant']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/scratch/slurm-7854814/ipykernel_385153/3328906579.py:20: DeprecationWarning: ref_level is deprecated and no longer has any effect. It will beremoved in a future release.\n",
      "/scratch/slurm-7854814/ipykernel_385153/3328906579.py:20: DeprecationWarning: design_factors is deprecated and will soon be removed.Please consider providing a formulaic formula using the design argumentinstead.\n",
      "Fitting size factors...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using None as control genes, passed at DeseqDataSet initialization\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... done in 0.26 seconds.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 1.44 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.11 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 1.60 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 1.28 seconds.\n",
      "\n",
      "Calculating cook's distance...\n",
      "... done in 0.65 seconds.\n",
      "\n",
      "Replacing 56 outlier genes.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 0.02 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 0.02 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 0.02 seconds.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 0.63 seconds.\n",
      "\n",
      "Fitting MAP LFCs...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: Group Resistant vs Naive\n",
      "                baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
      "FO538757.2      0.077367        0.287598  0.410237  0.701053  0.483270   \n",
      "AP006222.2      0.126624        0.150784  0.278099  0.542195  0.587684   \n",
      "RP4-669L17.10   0.003620        0.195583  3.709806  0.052720  0.957955   \n",
      "RP11-206L10.9   0.016739        0.196442  1.478265  0.132887  0.894283   \n",
      "FAM87B          0.003273        0.193681  3.709783  0.052208  0.958363   \n",
      "...                  ...             ...       ...       ...       ...   \n",
      "MT-CYB         21.041942       -0.037278  0.058619 -0.635933  0.524820   \n",
      "AL592183.1      0.024048        0.269137  0.850054  0.316611  0.751538   \n",
      "AC007325.4      0.034072        0.155855  0.710363  0.219402  0.826337   \n",
      "AC004556.1      0.089721       -1.410665  0.383780 -3.675714  0.000237   \n",
      "AC240274.1      0.072741        0.009621  0.527087  0.018253  0.985437   \n",
      "\n",
      "                   padj  \n",
      "FO538757.2          NaN  \n",
      "AP006222.2          NaN  \n",
      "RP4-669L17.10       NaN  \n",
      "RP11-206L10.9       NaN  \n",
      "FAM87B              NaN  \n",
      "...                 ...  \n",
      "MT-CYB         0.786037  \n",
      "AL592183.1          NaN  \n",
      "AC007325.4          NaN  \n",
      "AC004556.1          NaN  \n",
      "AC240274.1          NaN  \n",
      "\n",
      "[15464 rows x 6 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... done in 1.71 seconds.\n",
      "\n",
      "/scratch/slurm-7854814/ipykernel_385153/3328906579.py:20: DeprecationWarning: ref_level is deprecated and no longer has any effect. It will beremoved in a future release.\n",
      "/scratch/slurm-7854814/ipykernel_385153/3328906579.py:20: DeprecationWarning: design_factors is deprecated and will soon be removed.Please consider providing a formulaic formula using the design argumentinstead.\n",
      "Fitting size factors...\n",
      "... done in 0.04 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shrunk log2 fold change & Wald test p-value: Group[T.Resistant]\n",
      "                baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
      "FO538757.2      0.077367        0.146651  0.429453  0.701053  0.483270   \n",
      "AP006222.2      0.126624        0.011344  0.264949  0.542195  0.587684   \n",
      "RP4-669L17.10   0.003620       -0.001126  0.431099  0.052720  0.957955   \n",
      "RP11-206L10.9   0.016739       -0.014192  0.466888  0.132887  0.894283   \n",
      "FAM87B          0.003273       -0.008991  0.344451  0.052208  0.958363   \n",
      "...                  ...             ...       ...       ...       ...   \n",
      "MT-CYB         21.041942       -0.034624  0.059844 -0.635933  0.524820   \n",
      "AL592183.1      0.024048        0.050808  0.604738  0.316611  0.751538   \n",
      "AC007325.4      0.034072       -0.028075  0.400890  0.219402  0.826337   \n",
      "AC004556.1      0.089721       -7.953394  0.797497 -3.675714  0.000237   \n",
      "AC240274.1      0.072741       -0.214219  0.452171  0.018253  0.985437   \n",
      "\n",
      "                   padj  \n",
      "FO538757.2          NaN  \n",
      "AP006222.2          NaN  \n",
      "RP4-669L17.10       NaN  \n",
      "RP11-206L10.9       NaN  \n",
      "FAM87B              NaN  \n",
      "...                 ...  \n",
      "MT-CYB         0.786037  \n",
      "AL592183.1          NaN  \n",
      "AC007325.4          NaN  \n",
      "AC004556.1          NaN  \n",
      "AC240274.1          NaN  \n",
      "\n",
      "[15464 rows x 6 columns]\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "-------------------------------------------------------------------------------------\n",
      "\n",
      "HERE DC\n",
      "CELL TYPE:  DC ['Naive' 'Resistant']\n",
      "Using None as control genes, passed at DeseqDataSet initialization\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting dispersions...\n",
      "... done in 0.48 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.10 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 0.69 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 0.63 seconds.\n",
      "\n",
      "Calculating cook's distance...\n",
      "... done in 0.08 seconds.\n",
      "\n",
      "Replacing 39 outlier genes.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 0.01 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 0.01 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 0.01 seconds.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 0.67 seconds.\n",
      "\n",
      "Fitting MAP LFCs...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: Group Resistant vs Naive\n",
      "                baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
      "FO538757.2      0.082265       -0.026372  1.400646 -0.018829  0.984978   \n",
      "AP006222.2      0.100535       -0.069176  1.221144 -0.056648  0.954825   \n",
      "RP4-669L17.10   0.010220        0.006664  4.900878  0.001360  0.998915   \n",
      "RP11-206L10.9   0.013213        0.055207  4.275949  0.012911  0.989699   \n",
      "FAM87B          0.011609       -0.157388  4.851643 -0.032440  0.974121   \n",
      "...                  ...             ...       ...       ...       ...   \n",
      "MT-CYB         15.940078       -0.390111  0.147532 -2.644246  0.008187   \n",
      "AL592183.1      0.035720       -0.003724  2.905948 -0.001281  0.998978   \n",
      "AC007325.4      0.028423       -0.112320  3.142688 -0.035740  0.971489   \n",
      "AC004556.1      0.063247       -1.759061  1.929769 -0.911539  0.362011   \n",
      "AC240274.1      0.044981       -0.104128  1.910646 -0.054499  0.956538   \n",
      "\n",
      "                   padj  \n",
      "FO538757.2          NaN  \n",
      "AP006222.2          NaN  \n",
      "RP4-669L17.10       NaN  \n",
      "RP11-206L10.9       NaN  \n",
      "FAM87B              NaN  \n",
      "...                 ...  \n",
      "MT-CYB         0.059658  \n",
      "AL592183.1          NaN  \n",
      "AC007325.4          NaN  \n",
      "AC004556.1          NaN  \n",
      "AC240274.1          NaN  \n",
      "\n",
      "[15464 rows x 6 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... done in 0.73 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shrunk log2 fold change & Wald test p-value: Group[T.Resistant]\n",
      "                baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
      "FO538757.2      0.082265       -0.002275  0.762964 -0.018829  0.984978   \n",
      "AP006222.2      0.100535       -0.000247  0.720696 -0.056648  0.954825   \n",
      "RP4-669L17.10   0.010220        0.002458  1.449536  0.001360  0.998915   \n",
      "RP11-206L10.9   0.013213        0.004335  1.942730  0.012911  0.989699   \n",
      "FAM87B          0.011609       -0.014952  0.455286 -0.032440  0.974121   \n",
      "...                  ...             ...       ...       ...       ...   \n",
      "MT-CYB         15.940078       -0.334307  0.147104 -2.644246  0.008187   \n",
      "AL592183.1      0.035720       -0.000305  0.855772 -0.001281  0.998978   \n",
      "AC007325.4      0.028423       -0.015304  0.594334 -0.035740  0.971489   \n",
      "AC004556.1      0.063247       -8.751740  3.320695 -0.911539  0.362011   \n",
      "AC240274.1      0.044981       -0.020013  0.683063 -0.054499  0.956538   \n",
      "\n",
      "                   padj  \n",
      "FO538757.2          NaN  \n",
      "AP006222.2          NaN  \n",
      "RP4-669L17.10       NaN  \n",
      "RP11-206L10.9       NaN  \n",
      "FAM87B              NaN  \n",
      "...                 ...  \n",
      "MT-CYB         0.059658  \n",
      "AL592183.1          NaN  \n",
      "AC007325.4          NaN  \n",
      "AC004556.1          NaN  \n",
      "AC240274.1          NaN  \n",
      "\n",
      "[15464 rows x 6 columns]\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "-------------------------------------------------------------------------------------\n",
      "\n",
      "HERE Endothelial cells\n",
      "CELL TYPE:  Endothelial cells ['Naive' 'Resistant']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/scratch/slurm-7854814/ipykernel_385153/3328906579.py:20: DeprecationWarning: ref_level is deprecated and no longer has any effect. It will beremoved in a future release.\n",
      "/scratch/slurm-7854814/ipykernel_385153/3328906579.py:20: DeprecationWarning: design_factors is deprecated and will soon be removed.Please consider providing a formulaic formula using the design argumentinstead.\n",
      "Fitting size factors...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using None as control genes, passed at DeseqDataSet initialization\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... done in 0.25 seconds.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 0.85 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.11 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 1.58 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 1.21 seconds.\n",
      "\n",
      "Calculating cook's distance...\n",
      "... done in 0.59 seconds.\n",
      "\n",
      "Replacing 32 outlier genes.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 0.01 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 0.01 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 0.01 seconds.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 0.63 seconds.\n",
      "\n",
      "Fitting MAP LFCs...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: Group Resistant vs Naive\n",
      "                baseMean  log2FoldChange     lfcSE       stat        pvalue  \\\n",
      "FO538757.2      0.066491        0.861134  0.409488   2.102955  3.546969e-02   \n",
      "AP006222.2      0.085145        0.653732  0.347152   1.883129  5.968295e-02   \n",
      "RP4-669L17.10   0.004123        0.812777  2.597742   0.312878  7.543732e-01   \n",
      "RP11-206L10.9   0.012180        0.814035  1.397884   0.582334  5.603418e-01   \n",
      "FAM87B          0.001523        0.812882  3.361890   0.241793  8.089404e-01   \n",
      "...                  ...             ...       ...        ...           ...   \n",
      "MT-CYB         16.578296       -0.477356  0.035245 -13.544001  8.597604e-42   \n",
      "AL592183.1      0.040990        0.852229  0.634487   1.343178  1.792143e-01   \n",
      "AC007325.4      0.028887        0.751960  0.749859   1.002801  3.159569e-01   \n",
      "AC004556.1      0.043924        0.090817  0.464275   0.195611  8.449150e-01   \n",
      "AC240274.1      0.019594        0.807823  1.003262   0.805197  4.207062e-01   \n",
      "\n",
      "                       padj  \n",
      "FO538757.2              NaN  \n",
      "AP006222.2     1.012527e-01  \n",
      "RP4-669L17.10           NaN  \n",
      "RP11-206L10.9           NaN  \n",
      "FAM87B                  NaN  \n",
      "...                     ...  \n",
      "MT-CYB         5.530459e-40  \n",
      "AL592183.1              NaN  \n",
      "AC007325.4              NaN  \n",
      "AC004556.1              NaN  \n",
      "AC240274.1              NaN  \n",
      "\n",
      "[15464 rows x 6 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... done in 1.33 seconds.\n",
      "\n",
      "/scratch/slurm-7854814/ipykernel_385153/3328906579.py:20: DeprecationWarning: ref_level is deprecated and no longer has any effect. It will beremoved in a future release.\n",
      "/scratch/slurm-7854814/ipykernel_385153/3328906579.py:20: DeprecationWarning: design_factors is deprecated and will soon be removed.Please consider providing a formulaic formula using the design argumentinstead.\n",
      "Fitting size factors...\n",
      "... done in 0.01 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shrunk log2 fold change & Wald test p-value: Group[T.Resistant]\n",
      "                baseMean  log2FoldChange     lfcSE       stat        pvalue  \\\n",
      "FO538757.2      0.066491        0.391836  0.478948   2.102955  3.546969e-02   \n",
      "AP006222.2      0.085145       -0.040587  0.323153   1.883129  5.968295e-02   \n",
      "RP4-669L17.10   0.004123       -0.003120  0.594462   0.312878  7.543732e-01   \n",
      "RP11-206L10.9   0.012180        0.011723  0.617442   0.582334  5.603418e-01   \n",
      "FAM87B          0.001523       -0.026655  0.423126   0.241793  8.089404e-01   \n",
      "...                  ...             ...       ...        ...           ...   \n",
      "MT-CYB         16.578296       -0.474623  0.035503 -13.544001  8.597604e-42   \n",
      "AL592183.1      0.040990        0.210061  0.637490   1.343178  1.792143e-01   \n",
      "AC007325.4      0.028887       -0.026631  0.487559   1.002801  3.159569e-01   \n",
      "AC004556.1      0.043924       -6.360706  0.921248   0.195611  8.449150e-01   \n",
      "AC240274.1      0.019594       -0.006750  0.553882   0.805197  4.207062e-01   \n",
      "\n",
      "                       padj  \n",
      "FO538757.2              NaN  \n",
      "AP006222.2     1.012527e-01  \n",
      "RP4-669L17.10           NaN  \n",
      "RP11-206L10.9           NaN  \n",
      "FAM87B                  NaN  \n",
      "...                     ...  \n",
      "MT-CYB         5.530459e-40  \n",
      "AL592183.1              NaN  \n",
      "AC007325.4              NaN  \n",
      "AC004556.1              NaN  \n",
      "AC240274.1              NaN  \n",
      "\n",
      "[15464 rows x 6 columns]\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "-------------------------------------------------------------------------------------\n",
      "\n",
      "HERE Epithelial cells\n",
      "CELL TYPE:  Epithelial cells ['Naive' 'Resistant']\n",
      "Using None as control genes, passed at DeseqDataSet initialization\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting dispersions...\n",
      "... done in 0.41 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.08 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 0.61 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 0.53 seconds.\n",
      "\n",
      "Calculating cook's distance...\n",
      "... done in 0.01 seconds.\n",
      "\n",
      "Replacing 25 outlier genes.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 0.01 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 0.01 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 0.01 seconds.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 0.56 seconds.\n",
      "\n",
      "Fitting MAP LFCs...\n",
      "/ihome/rgottschalk/cil8/envs/rnaseq/lib/python3.13/site-packages/pydeseq2/utils.py:1088: RuntimeWarning: overflow encountered in exp\n",
      "  counts - (counts + size) / (1 + size * np.exp(-xbeta - offset))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: Group Resistant vs Naive\n",
      "                baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
      "FO538757.2      0.102327       -0.457108  2.956045 -0.154635  0.877109   \n",
      "AP006222.2      0.181390       -0.863947  0.981804 -0.879959  0.378881   \n",
      "RP4-669L17.10   0.000000             NaN       NaN       NaN       NaN   \n",
      "RP11-206L10.9   0.194018       -0.400935  3.338419 -0.120097  0.904406   \n",
      "FAM87B          0.000000             NaN       NaN       NaN       NaN   \n",
      "...                  ...             ...       ...       ...       ...   \n",
      "MT-CYB         29.692949       -1.381261  0.300476 -4.596900  0.000004   \n",
      "AL592183.1      0.000000             NaN       NaN       NaN       NaN   \n",
      "AC007325.4      0.032608       -0.278153  3.335468 -0.083393  0.933539   \n",
      "AC004556.1      0.060287       -1.221625  2.422633 -0.504255  0.614082   \n",
      "AC240274.1      0.112279       -0.833132  3.317112 -0.251162  0.801689   \n",
      "\n",
      "                   padj  \n",
      "FO538757.2          NaN  \n",
      "AP006222.2          NaN  \n",
      "RP4-669L17.10       NaN  \n",
      "RP11-206L10.9       NaN  \n",
      "FAM87B              NaN  \n",
      "...                 ...  \n",
      "MT-CYB         0.000143  \n",
      "AL592183.1          NaN  \n",
      "AC007325.4          NaN  \n",
      "AC004556.1          NaN  \n",
      "AC240274.1          NaN  \n",
      "\n",
      "[15464 rows x 6 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/ihome/rgottschalk/cil8/envs/rnaseq/lib/python3.13/site-packages/pydeseq2/utils.py:1088: RuntimeWarning: overflow encountered in exp\n",
      "  counts - (counts + size) / (1 + size * np.exp(-xbeta - offset))\n",
      "/ihome/rgottschalk/cil8/envs/rnaseq/lib/python3.13/site-packages/pydeseq2/utils.py:1088: RuntimeWarning: overflow encountered in exp\n",
      "  counts - (counts + size) / (1 + size * np.exp(-xbeta - offset))\n",
      "/ihome/rgottschalk/cil8/envs/rnaseq/lib/python3.13/site-packages/pydeseq2/utils.py:1088: RuntimeWarning: overflow encountered in exp\n",
      "  counts - (counts + size) / (1 + size * np.exp(-xbeta - offset))\n",
      "/ihome/rgottschalk/cil8/envs/rnaseq/lib/python3.13/site-packages/pydeseq2/utils.py:1088: RuntimeWarning: overflow encountered in exp\n",
      "  counts - (counts + size) / (1 + size * np.exp(-xbeta - offset))\n",
      "/ihome/rgottschalk/cil8/envs/rnaseq/lib/python3.13/site-packages/pydeseq2/utils.py:1088: RuntimeWarning: overflow encountered in exp\n",
      "  counts - (counts + size) / (1 + size * np.exp(-xbeta - offset))\n",
      "... done in 0.95 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shrunk log2 fold change & Wald test p-value: Group[T.Resistant]\n",
      "                baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
      "FO538757.2      0.102327   -1.964772e-07  0.001530 -0.154635  0.877109   \n",
      "AP006222.2      0.181390   -1.282600e-06  0.001722 -0.879959  0.378881   \n",
      "RP4-669L17.10   0.000000             NaN       NaN       NaN       NaN   \n",
      "RP11-206L10.9   0.194018   -3.445105e-07  0.001375 -0.120097  0.904406   \n",
      "FAM87B          0.000000             NaN       NaN       NaN       NaN   \n",
      "...                  ...             ...       ...       ...       ...   \n",
      "MT-CYB         29.692949   -2.816389e-05  0.001361 -4.596900  0.000004   \n",
      "AL592183.1      0.000000             NaN       NaN       NaN       NaN   \n",
      "AC007325.4      0.032608    6.102593e-09  0.002021 -0.083393  0.933539   \n",
      "AC004556.1      0.060287   -7.198409e-07  0.001138 -0.504255  0.614082   \n",
      "AC240274.1      0.112279   -4.293589e-07  0.001118 -0.251162  0.801689   \n",
      "\n",
      "                   padj  \n",
      "FO538757.2          NaN  \n",
      "AP006222.2          NaN  \n",
      "RP4-669L17.10       NaN  \n",
      "RP11-206L10.9       NaN  \n",
      "FAM87B              NaN  \n",
      "...                 ...  \n",
      "MT-CYB         0.000143  \n",
      "AL592183.1          NaN  \n",
      "AC007325.4          NaN  \n",
      "AC004556.1          NaN  \n",
      "AC240274.1          NaN  \n",
      "\n",
      "[15464 rows x 6 columns]\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "-------------------------------------------------------------------------------------\n",
      "\n",
      "HERE T cells\n",
      "CELL TYPE:  T cells ['Naive' 'Resistant']\n",
      "Using None as control genes, passed at DeseqDataSet initialization\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/scratch/slurm-7854814/ipykernel_385153/3328906579.py:20: DeprecationWarning: ref_level is deprecated and no longer has any effect. It will beremoved in a future release.\n",
      "/scratch/slurm-7854814/ipykernel_385153/3328906579.py:20: DeprecationWarning: design_factors is deprecated and will soon be removed.Please consider providing a formulaic formula using the design argumentinstead.\n",
      "Fitting size factors...\n",
      "... done in 0.08 seconds.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 0.56 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.10 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 0.97 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 0.64 seconds.\n",
      "\n",
      "Calculating cook's distance...\n",
      "... done in 0.18 seconds.\n",
      "\n",
      "Replacing 50 outlier genes.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 0.01 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 0.01 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 0.01 seconds.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 0.56 seconds.\n",
      "\n",
      "Fitting MAP LFCs...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: Group Resistant vs Naive\n",
      "               baseMean  log2FoldChange     lfcSE      stat    pvalue     padj\n",
      "FO538757.2     0.049730        0.822542  2.045470  0.402129  0.687589      NaN\n",
      "AP006222.2     0.021730        0.770789  2.987400  0.258013  0.796397      NaN\n",
      "RP4-669L17.10  0.002562        0.719016  5.606360  0.128250  0.897951      NaN\n",
      "RP11-206L10.9  0.012362        0.740579  4.742109  0.156171  0.875898      NaN\n",
      "FAM87B         0.001148        0.714709  5.606365  0.127482  0.898559      NaN\n",
      "...                 ...             ...       ...       ...       ...      ...\n",
      "MT-CYB         9.756633       -0.045537  0.138432 -0.328950  0.742193  0.84355\n",
      "AL592183.1     0.028288        0.783729  2.609314  0.300358  0.763904      NaN\n",
      "AC007325.4     0.011606        0.633927  4.065642  0.155923  0.876094      NaN\n",
      "AC004556.1     0.009667        0.099981  4.092986  0.024427  0.980512      NaN\n",
      "AC240274.1     0.008019        0.676456  5.346468  0.126524  0.899317      NaN\n",
      "\n",
      "[15464 rows x 6 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/ihome/rgottschalk/cil8/envs/rnaseq/lib/python3.13/site-packages/pydeseq2/utils.py:1088: RuntimeWarning: overflow encountered in exp\n",
      "  counts - (counts + size) / (1 + size * np.exp(-xbeta - offset))\n",
      "... done in 0.97 seconds.\n",
      "\n",
      "/scratch/slurm-7854814/ipykernel_385153/3328906579.py:20: DeprecationWarning: ref_level is deprecated and no longer has any effect. It will beremoved in a future release.\n",
      "/scratch/slurm-7854814/ipykernel_385153/3328906579.py:20: DeprecationWarning: design_factors is deprecated and will soon be removed.Please consider providing a formulaic formula using the design argumentinstead.\n",
      "Fitting size factors...\n",
      "... done in 0.01 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shrunk log2 fold change & Wald test p-value: Group[T.Resistant]\n",
      "               baseMean  log2FoldChange     lfcSE      stat    pvalue     padj\n",
      "FO538757.2     0.049730        0.014159  1.780373  0.402129  0.687589      NaN\n",
      "AP006222.2     0.021730        0.006189  1.961612  0.258013  0.796397      NaN\n",
      "RP4-669L17.10  0.002562        0.000825  1.442778  0.128250  0.897951      NaN\n",
      "RP11-206L10.9  0.012362        0.002122  2.266399  0.156171  0.875898      NaN\n",
      "FAM87B         0.001148        0.000418  1.128413  0.127482  0.898559      NaN\n",
      "...                 ...             ...       ...       ...       ...      ...\n",
      "MT-CYB         9.756633       -0.031486  0.136592 -0.328950  0.742193  0.84355\n",
      "AL592183.1     0.028288        0.008351  1.911559  0.300358  0.763904      NaN\n",
      "AC007325.4     0.011606       -0.005400  0.512408  0.155923  0.876094      NaN\n",
      "AC004556.1     0.009667       -0.042214  0.236338  0.024427  0.980512      NaN\n",
      "AC240274.1     0.008019       -0.002249  0.576749  0.126524  0.899317      NaN\n",
      "\n",
      "[15464 rows x 6 columns]\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "-------------------------------------------------------------------------------------\n",
      "\n",
      "HERE ILC\n",
      "CELL TYPE:  ILC ['Naive' 'Resistant']\n",
      "Using None as control genes, passed at DeseqDataSet initialization\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting dispersions...\n",
      "... done in 0.43 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.08 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 0.62 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 0.57 seconds.\n",
      "\n",
      "Calculating cook's distance...\n",
      "... done in 0.03 seconds.\n",
      "\n",
      "Replacing 47 outlier genes.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 0.01 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 0.01 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 0.01 seconds.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 0.52 seconds.\n",
      "\n",
      "Fitting MAP LFCs...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: Group Resistant vs Naive\n",
      "                baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
      "FO538757.2      0.043387        0.341331  3.461772  0.098600  0.921456   \n",
      "AP006222.2      0.032893        0.039225  3.877585  0.010116  0.991929   \n",
      "RP4-669L17.10   0.000000             NaN       NaN       NaN       NaN   \n",
      "RP11-206L10.9   0.021635        0.250291  5.222568  0.047925  0.961776   \n",
      "FAM87B          0.000000             NaN       NaN       NaN       NaN   \n",
      "...                  ...             ...       ...       ...       ...   \n",
      "MT-CYB         16.822940        0.061010  0.319832  0.190757  0.848716   \n",
      "AL592183.1      0.050585        0.303499  3.898432  0.077851  0.937946   \n",
      "AC007325.4      0.022922        0.229536  5.222699  0.043950  0.964945   \n",
      "AC004556.1      0.004783       -0.078411  5.206840 -0.015059  0.987985   \n",
      "AC240274.1      0.059818        0.327367  4.511393  0.072564  0.942153   \n",
      "\n",
      "                   padj  \n",
      "FO538757.2          NaN  \n",
      "AP006222.2          NaN  \n",
      "RP4-669L17.10       NaN  \n",
      "RP11-206L10.9       NaN  \n",
      "FAM87B              NaN  \n",
      "...                 ...  \n",
      "MT-CYB         0.964068  \n",
      "AL592183.1          NaN  \n",
      "AC007325.4          NaN  \n",
      "AC004556.1          NaN  \n",
      "AC240274.1          NaN  \n",
      "\n",
      "[15464 rows x 6 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... done in 0.75 seconds.\n",
      "\n",
      "/scratch/slurm-7854814/ipykernel_385153/3328906579.py:20: DeprecationWarning: ref_level is deprecated and no longer has any effect. It will beremoved in a future release.\n",
      "/scratch/slurm-7854814/ipykernel_385153/3328906579.py:20: DeprecationWarning: design_factors is deprecated and will soon be removed.Please consider providing a formulaic formula using the design argumentinstead.\n",
      "Fitting size factors...\n",
      "... done in 0.02 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shrunk log2 fold change & Wald test p-value: Group[T.Resistant]\n",
      "                baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
      "FO538757.2      0.043387        0.003310  1.230527  0.098600  0.921456   \n",
      "AP006222.2      0.032893       -0.000852  0.548893  0.010116  0.991929   \n",
      "RP4-669L17.10   0.000000             NaN       NaN       NaN       NaN   \n",
      "RP11-206L10.9   0.021635        0.001189  1.337273  0.047925  0.961776   \n",
      "FAM87B          0.000000             NaN       NaN       NaN       NaN   \n",
      "...                  ...             ...       ...       ...       ...   \n",
      "MT-CYB         16.822940        0.014423  0.297027  0.190757  0.848716   \n",
      "AL592183.1      0.050585        0.002700  1.349245  0.077851  0.937946   \n",
      "AC007325.4      0.022922        0.001167  1.267130  0.043950  0.964945   \n",
      "AC004556.1      0.004783       -0.009877  0.248485 -0.015059  0.987985   \n",
      "AC240274.1      0.059818        0.002101  1.646228  0.072564  0.942153   \n",
      "\n",
      "                   padj  \n",
      "FO538757.2          NaN  \n",
      "AP006222.2          NaN  \n",
      "RP4-669L17.10       NaN  \n",
      "RP11-206L10.9       NaN  \n",
      "FAM87B              NaN  \n",
      "...                 ...  \n",
      "MT-CYB         0.964068  \n",
      "AL592183.1          NaN  \n",
      "AC007325.4          NaN  \n",
      "AC004556.1          NaN  \n",
      "AC240274.1          NaN  \n",
      "\n",
      "[15464 rows x 6 columns]\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "-------------------------------------------------------------------------------------\n",
      "\n",
      "HERE B cells\n",
      "CELL TYPE:  B cells ['Naive' 'Resistant']\n",
      "Using None as control genes, passed at DeseqDataSet initialization\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting dispersions...\n",
      "... done in 0.43 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.09 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 0.60 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 0.56 seconds.\n",
      "\n",
      "Calculating cook's distance...\n",
      "... done in 0.04 seconds.\n",
      "\n",
      "Replacing 29 outlier genes.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 0.01 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 0.01 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 0.01 seconds.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 0.52 seconds.\n",
      "\n",
      "Fitting MAP LFCs...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: Group Resistant vs Naive\n",
      "                baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
      "FO538757.2      0.028776        0.824975  5.029180  0.164038  0.869701   \n",
      "AP006222.2      0.057320       -0.054644  2.528680 -0.021610  0.982759   \n",
      "RP4-669L17.10   0.006075        0.754460  6.216707  0.121360  0.903406   \n",
      "RP11-206L10.9   0.009659        0.522548  6.204435  0.084222  0.932880   \n",
      "FAM87B          0.002557        0.498455  6.204518  0.080337  0.935969   \n",
      "...                  ...             ...       ...       ...       ...   \n",
      "MT-CYB         11.464372       -0.415449  0.278997 -1.489076  0.136467   \n",
      "AL592183.1      0.038131        0.847963  5.033978  0.168448  0.866231   \n",
      "AC007325.4      0.005944        0.520408  6.204442  0.083877  0.933154   \n",
      "AC004556.1      0.011990        0.008937  6.185019  0.001445  0.998847   \n",
      "AC240274.1      0.032120        0.575354  5.014122  0.114747  0.908646   \n",
      "\n",
      "                   padj  \n",
      "FO538757.2          NaN  \n",
      "AP006222.2          NaN  \n",
      "RP4-669L17.10       NaN  \n",
      "RP11-206L10.9       NaN  \n",
      "FAM87B              NaN  \n",
      "...                 ...  \n",
      "MT-CYB         0.484014  \n",
      "AL592183.1          NaN  \n",
      "AC007325.4          NaN  \n",
      "AC004556.1          NaN  \n",
      "AC240274.1          NaN  \n",
      "\n",
      "[15464 rows x 6 columns]\n",
      "Shrunk log2 fold change & Wald test p-value: Group[T.Resistant]\n",
      "                baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
      "FO538757.2      0.028776        0.002125  2.272513  0.164038  0.869701   \n",
      "AP006222.2      0.057320       -0.018580  0.517555 -0.021610  0.982759   \n",
      "RP4-669L17.10   0.006075        0.000528  1.358364  0.121360  0.903406   \n",
      "RP11-206L10.9   0.009659       -0.004066  0.492782  0.084222  0.932880   \n",
      "FAM87B          0.002557       -0.013778  0.304422  0.080337  0.935969   \n",
      "...                  ...             ...       ...       ...       ...   \n",
      "MT-CYB         11.464372       -0.198533  0.263127 -1.489076  0.136467   \n",
      "AL592183.1      0.038131        0.002248  2.542223  0.168448  0.866231   \n",
      "AC007325.4      0.005944       -0.009670  0.388551  0.083877  0.933154   \n",
      "AC004556.1      0.011990       -0.023007  0.259997  0.001445  0.998847   \n",
      "AC240274.1      0.032120       -0.002380  0.665172  0.114747  0.908646   \n",
      "\n",
      "                   padj  \n",
      "FO538757.2          NaN  \n",
      "AP006222.2          NaN  \n",
      "RP4-669L17.10       NaN  \n",
      "RP11-206L10.9       NaN  \n",
      "FAM87B              NaN  \n",
      "...                 ...  \n",
      "MT-CYB         0.484014  \n",
      "AL592183.1          NaN  \n",
      "AC007325.4          NaN  \n",
      "AC004556.1          NaN  \n",
      "AC240274.1          NaN  \n",
      "\n",
      "[15464 rows x 6 columns]\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "-------------------------------------------------------------------------------------\n",
      "\n",
      "HERE Monocytes\n",
      "CELL TYPE:  Monocytes ['Resistant']\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "-------------------------------------------------------------------------------------\n",
      "\n",
      "HERE Plasma cells\n",
      "CELL TYPE:  Plasma cells ['Resistant']\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "-------------------------------------------------------------------------------------\n",
      "\n",
      "HERE pDC\n",
      "CELL TYPE:  pDC ['Resistant']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... done in 0.78 seconds.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "condition_key = \"Group\"\n",
    "groupby = \"cell_type\"\n",
    "dea_results = {}\n",
    "quiet = True\n",
    "\n",
    "for cell_group in adata.obs[groupby].unique():\n",
    "    # Select cell profiles\n",
    "    ctdata = adata[adata.obs[groupby] == cell_group].copy()\n",
    "    print(\"\\n\\n\\n\\n\\n-------------------------------------------------------------------------------------\\n\\nHERE\", cell_group); \n",
    "    print(\"CELL TYPE: \", cell_group, ctdata.obs[\"Group\"].unique())\n",
    "\n",
    "    # Check to see if cell type is present in both Groups\n",
    "    if(len(np.unique(ctdata.obs[\"Group\"])) < 2):\n",
    "        # remove it from the dataset and continue \n",
    "        adata = adata[adata.obs[groupby] != cell_group]\n",
    "        continue\n",
    "        \n",
    "    # Build DESeq2 object\n",
    "    # NOTE: this data is actually paired, so one could consider fitting the patient label as a confounder\n",
    "    dds = DeseqDataSet(\n",
    "    \tadata=ctdata,\n",
    "    \tdesign_factors=condition_key,\n",
    "    \tref_level=[condition_key, 'Naive'], # set control as reference\n",
    "    \trefit_cooks=True,\n",
    "    )\n",
    "\n",
    "    # Compute LFCs\n",
    "    dds.deseq2()\n",
    "    # Contrast between stim and ctrl\n",
    "    stat_res = DeseqStats(dds, contrast=[condition_key, 'Resistant', 'Naive'])\n",
    "    # Compute Wald test\n",
    "    stat_res.summary()\n",
    "    # Shrink LFCs\n",
    "    stat_res.lfc_shrink(coeff='Group[T.Resistant]') # {condition_key}_cond_vs_ref\n",
    "    #stat_res.lfc_shrink(coeff='condition_Resistant_vs_Naive') # {condition_key}_cond_vs_ref\n",
    "\n",
    "    dea_results[cell_group] = stat_res.results_df;\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "3ee242f2-c597-4412-b131-8e2ba8dba0d9",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_type</th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>FO538757.2</th>\n",
       "      <td>Fibroblasts</td>\n",
       "      <td>0.124881</td>\n",
       "      <td>0.566587</td>\n",
       "      <td>0.280376</td>\n",
       "      <td>2.924109</td>\n",
       "      <td>0.003454</td>\n",
       "      <td>0.009842</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AP006222.2</th>\n",
       "      <td>Fibroblasts</td>\n",
       "      <td>0.184819</td>\n",
       "      <td>-0.129691</td>\n",
       "      <td>0.169100</td>\n",
       "      <td>1.891515</td>\n",
       "      <td>0.058556</td>\n",
       "      <td>0.110536</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RP4-669L17.10</th>\n",
       "      <td>Fibroblasts</td>\n",
       "      <td>0.004709</td>\n",
       "      <td>-0.001118</td>\n",
       "      <td>0.875506</td>\n",
       "      <td>0.345265</td>\n",
       "      <td>0.729895</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RP11-206L10.9</th>\n",
       "      <td>Fibroblasts</td>\n",
       "      <td>0.027324</td>\n",
       "      <td>0.046824</td>\n",
       "      <td>0.611874</td>\n",
       "      <td>1.028176</td>\n",
       "      <td>0.303867</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM87B</th>\n",
       "      <td>Fibroblasts</td>\n",
       "      <td>0.010583</td>\n",
       "      <td>-0.224822</td>\n",
       "      <td>0.633389</td>\n",
       "      <td>0.576261</td>\n",
       "      <td>0.564439</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 cell_type  baseMean  log2FoldChange     lfcSE      stat  \\\n",
       "index                                                                      \n",
       "FO538757.2     Fibroblasts  0.124881        0.566587  0.280376  2.924109   \n",
       "AP006222.2     Fibroblasts  0.184819       -0.129691  0.169100  1.891515   \n",
       "RP4-669L17.10  Fibroblasts  0.004709       -0.001118  0.875506  0.345265   \n",
       "RP11-206L10.9  Fibroblasts  0.027324        0.046824  0.611874  1.028176   \n",
       "FAM87B         Fibroblasts  0.010583       -0.224822  0.633389  0.576261   \n",
       "\n",
       "                 pvalue      padj  \n",
       "index                              \n",
       "FO538757.2     0.003454  0.009842  \n",
       "AP006222.2     0.058556  0.110536  \n",
       "RP4-669L17.10  0.729895       NaN  \n",
       "RP11-206L10.9  0.303867       NaN  \n",
       "FAM87B         0.564439       NaN  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# concat results across cell types\n",
    "dea_df = pd.concat(dea_results)\n",
    "dea_df = dea_df.reset_index().rename(columns={'level_0': groupby,'level_1':'index'}).set_index('index')\n",
    "dea_df.head()\n",
    "\n",
    "# Shrink LFCs\n",
    "#stat_res.lfc_shrink(coeff='Group[T.Resistant]') # {condition_key}_cond_vs_ref\n",
    "#stat_res.results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "6d781ba3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# PyDeseq Seems to intrdoce NAs for some p-values\n",
    "# NOTE: there sometimes some NaN being introduced, best to double check that, in this case it's only for a single gene, but it might be a problem.\n",
    "len(dea_df[dea_df.isna().any(axis=1)])\n",
    "dea_df.to_csv(\"deseq_results.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "1f4e0805",
   "metadata": {},
   "outputs": [],
   "source": [
    "#adata = adata[adata.obs[condition_key]=='Resistant'].copy()\n",
    "sc.pp.normalize_total(adata, target_sum=1e4)\n",
    "sc.pp.log1p(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "2f37090e-15c4-4c53-b766-da6298458dae",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using resource `consensus`.\n",
      "Using `.X`!\n",
      "Converting to sparse csr matrix!\n",
      "0.39 of entities in the resource are missing from the data.\n"
     ]
    }
   ],
   "source": [
    "lr_res = li.multi.df_to_lr(adata,\n",
    "                           dea_df=dea_df,\n",
    "                           resource_name='consensus', # NOTE: uses HUMAN gene symbols!\n",
    "                           expr_prop=0.1, # calculated for adata as passed - used to filter interactions\n",
    "                           groupby=\"cell_type\",\n",
    "                           stat_keys=['stat', 'pvalue', 'padj'],\n",
    "                           use_raw=False,\n",
    "                           complex_col='stat', # NOTE: we use the Wald Stat to deal with complexes\n",
    "                           verbose=True,\n",
    "                           return_all_lrs=False,\n",
    "                           )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "6ba8726c-e044-47b1-808a-201f5277cf3d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ligand</th>\n",
       "      <th>receptor</th>\n",
       "      <th>ligand_complex</th>\n",
       "      <th>receptor_complex</th>\n",
       "      <th>source</th>\n",
       "      <th>ligand_stat</th>\n",
       "      <th>ligand_pvalue</th>\n",
       "      <th>ligand_padj</th>\n",
       "      <th>ligand_expr</th>\n",
       "      <th>ligand_props</th>\n",
       "      <th>...</th>\n",
       "      <th>receptor_pvalue</th>\n",
       "      <th>receptor_padj</th>\n",
       "      <th>receptor_expr</th>\n",
       "      <th>receptor_props</th>\n",
       "      <th>interaction_stat</th>\n",
       "      <th>interaction_pvalue</th>\n",
       "      <th>interaction_padj</th>\n",
       "      <th>interaction_expr</th>\n",
       "      <th>interaction_props</th>\n",
       "      <th>interaction</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4104</th>\n",
       "      <td>VIM</td>\n",
       "      <td>CD44</td>\n",
       "      <td>VIM</td>\n",
       "      <td>CD44</td>\n",
       "      <td>Endothelial cells</td>\n",
       "      <td>-34.265628</td>\n",
       "      <td>2.552207e-257</td>\n",
       "      <td>1.283528e-254</td>\n",
       "      <td>3.681040</td>\n",
       "      <td>0.999127</td>\n",
       "      <td>...</td>\n",
       "      <td>2.489514e-09</td>\n",
       "      <td>2.646970e-08</td>\n",
       "      <td>1.337293</td>\n",
       "      <td>0.762220</td>\n",
       "      <td>-20.113885</td>\n",
       "      <td>1.244757e-09</td>\n",
       "      <td>1.323485e-08</td>\n",
       "      <td>2.509166</td>\n",
       "      <td>0.880674</td>\n",
       "      <td>VIM^CD44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8424</th>\n",
       "      <td>VIM</td>\n",
       "      <td>CD44</td>\n",
       "      <td>VIM</td>\n",
       "      <td>CD44</td>\n",
       "      <td>Fibroblasts</td>\n",
       "      <td>-32.616964</td>\n",
       "      <td>2.357450e-233</td>\n",
       "      <td>9.790353e-231</td>\n",
       "      <td>3.606308</td>\n",
       "      <td>0.998240</td>\n",
       "      <td>...</td>\n",
       "      <td>2.489514e-09</td>\n",
       "      <td>2.646970e-08</td>\n",
       "      <td>1.337293</td>\n",
       "      <td>0.762220</td>\n",
       "      <td>-19.289554</td>\n",
       "      <td>1.244757e-09</td>\n",
       "      <td>1.323485e-08</td>\n",
       "      <td>2.471801</td>\n",
       "      <td>0.880230</td>\n",
       "      <td>VIM^CD44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8329</th>\n",
       "      <td>CXCL12</td>\n",
       "      <td>CXCR4</td>\n",
       "      <td>CXCL12</td>\n",
       "      <td>CXCR4</td>\n",
       "      <td>Fibroblasts</td>\n",
       "      <td>30.101009</td>\n",
       "      <td>4.700240e-199</td>\n",
       "      <td>1.229026e-196</td>\n",
       "      <td>2.238536</td>\n",
       "      <td>0.780211</td>\n",
       "      <td>...</td>\n",
       "      <td>3.563692e-11</td>\n",
       "      <td>4.462986e-10</td>\n",
       "      <td>0.610545</td>\n",
       "      <td>0.376968</td>\n",
       "      <td>18.361089</td>\n",
       "      <td>1.781846e-11</td>\n",
       "      <td>2.231493e-10</td>\n",
       "      <td>1.424540</td>\n",
       "      <td>0.578589</td>\n",
       "      <td>CXCL12^CXCR4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8291</th>\n",
       "      <td>APOE</td>\n",
       "      <td>SDC2</td>\n",
       "      <td>APOE</td>\n",
       "      <td>SDC2</td>\n",
       "      <td>Fibroblasts</td>\n",
       "      <td>33.165903</td>\n",
       "      <td>3.340814e-241</td>\n",
       "      <td>1.814319e-238</td>\n",
       "      <td>2.378408</td>\n",
       "      <td>0.738561</td>\n",
       "      <td>...</td>\n",
       "      <td>3.169624e-03</td>\n",
       "      <td>1.353437e-02</td>\n",
       "      <td>0.227950</td>\n",
       "      <td>0.146645</td>\n",
       "      <td>18.058347</td>\n",
       "      <td>1.584812e-03</td>\n",
       "      <td>6.767183e-03</td>\n",
       "      <td>1.303179</td>\n",
       "      <td>0.442603</td>\n",
       "      <td>APOE^SDC2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8443</th>\n",
       "      <td>APOE</td>\n",
       "      <td>ABCA1</td>\n",
       "      <td>APOE</td>\n",
       "      <td>ABCA1</td>\n",
       "      <td>Fibroblasts</td>\n",
       "      <td>33.165903</td>\n",
       "      <td>3.340814e-241</td>\n",
       "      <td>1.814319e-238</td>\n",
       "      <td>2.378408</td>\n",
       "      <td>0.738561</td>\n",
       "      <td>...</td>\n",
       "      <td>2.307151e-01</td>\n",
       "      <td>4.837241e-01</td>\n",
       "      <td>0.494338</td>\n",
       "      <td>0.356669</td>\n",
       "      <td>17.182211</td>\n",
       "      <td>1.153576e-01</td>\n",
       "      <td>2.418620e-01</td>\n",
       "      <td>1.436373</td>\n",
       "      <td>0.547615</td>\n",
       "      <td>APOE^ABCA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8382</th>\n",
       "      <td>APOE</td>\n",
       "      <td>SORL1</td>\n",
       "      <td>APOE</td>\n",
       "      <td>SORL1</td>\n",
       "      <td>Fibroblasts</td>\n",
       "      <td>33.165903</td>\n",
       "      <td>3.340814e-241</td>\n",
       "      <td>1.814319e-238</td>\n",
       "      <td>2.378408</td>\n",
       "      <td>0.738561</td>\n",
       "      <td>...</td>\n",
       "      <td>5.311427e-01</td>\n",
       "      <td>7.890612e-01</td>\n",
       "      <td>0.288026</td>\n",
       "      <td>0.225352</td>\n",
       "      <td>16.896083</td>\n",
       "      <td>2.655714e-01</td>\n",
       "      <td>3.945306e-01</td>\n",
       "      <td>1.333217</td>\n",
       "      <td>0.481956</td>\n",
       "      <td>APOE^SORL1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8368</th>\n",
       "      <td>ANGPTL2</td>\n",
       "      <td>ITGB1</td>\n",
       "      <td>ANGPTL2</td>\n",
       "      <td>ITGA5_ITGB1</td>\n",
       "      <td>Fibroblasts</td>\n",
       "      <td>-32.630403</td>\n",
       "      <td>1.520047e-233</td>\n",
       "      <td>6.707206e-231</td>\n",
       "      <td>0.715059</td>\n",
       "      <td>0.482988</td>\n",
       "      <td>...</td>\n",
       "      <td>4.684623e-01</td>\n",
       "      <td>7.411197e-01</td>\n",
       "      <td>0.629548</td>\n",
       "      <td>0.463546</td>\n",
       "      <td>-16.677693</td>\n",
       "      <td>2.342312e-01</td>\n",
       "      <td>3.705599e-01</td>\n",
       "      <td>0.672304</td>\n",
       "      <td>0.473267</td>\n",
       "      <td>ANGPTL2^ITGA5_ITGB1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8327</th>\n",
       "      <td>CXCL12</td>\n",
       "      <td>ITGAV</td>\n",
       "      <td>CXCL12</td>\n",
       "      <td>ITGAV</td>\n",
       "      <td>Fibroblasts</td>\n",
       "      <td>30.101009</td>\n",
       "      <td>4.700240e-199</td>\n",
       "      <td>1.229026e-196</td>\n",
       "      <td>2.238536</td>\n",
       "      <td>0.780211</td>\n",
       "      <td>...</td>\n",
       "      <td>6.259711e-03</td>\n",
       "      <td>2.474708e-02</td>\n",
       "      <td>0.245828</td>\n",
       "      <td>0.198840</td>\n",
       "      <td>16.417433</td>\n",
       "      <td>3.129855e-03</td>\n",
       "      <td>1.237354e-02</td>\n",
       "      <td>1.242182</td>\n",
       "      <td>0.489526</td>\n",
       "      <td>CXCL12^ITGAV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8447</th>\n",
       "      <td>TIMP2</td>\n",
       "      <td>CD44</td>\n",
       "      <td>TIMP2</td>\n",
       "      <td>CD44</td>\n",
       "      <td>Fibroblasts</td>\n",
       "      <td>-26.735568</td>\n",
       "      <td>1.817332e-157</td>\n",
       "      <td>3.054849e-155</td>\n",
       "      <td>1.093515</td>\n",
       "      <td>0.704145</td>\n",
       "      <td>...</td>\n",
       "      <td>2.489514e-09</td>\n",
       "      <td>2.646970e-08</td>\n",
       "      <td>1.337293</td>\n",
       "      <td>0.762220</td>\n",
       "      <td>-16.348856</td>\n",
       "      <td>1.244757e-09</td>\n",
       "      <td>1.323485e-08</td>\n",
       "      <td>1.215404</td>\n",
       "      <td>0.733183</td>\n",
       "      <td>TIMP2^CD44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8369</th>\n",
       "      <td>ANGPTL2</td>\n",
       "      <td>TLR4</td>\n",
       "      <td>ANGPTL2</td>\n",
       "      <td>TLR4</td>\n",
       "      <td>Fibroblasts</td>\n",
       "      <td>-32.630403</td>\n",
       "      <td>1.520047e-233</td>\n",
       "      <td>6.707206e-231</td>\n",
       "      <td>0.715059</td>\n",
       "      <td>0.482988</td>\n",
       "      <td>...</td>\n",
       "      <td>9.549634e-01</td>\n",
       "      <td>9.848294e-01</td>\n",
       "      <td>0.195188</td>\n",
       "      <td>0.163629</td>\n",
       "      <td>-16.286964</td>\n",
       "      <td>4.774817e-01</td>\n",
       "      <td>4.924147e-01</td>\n",
       "      <td>0.455124</td>\n",
       "      <td>0.323308</td>\n",
       "      <td>ANGPTL2^TLR4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows Ã— 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       ligand receptor ligand_complex receptor_complex             source  \\\n",
       "4104      VIM     CD44            VIM             CD44  Endothelial cells   \n",
       "8424      VIM     CD44            VIM             CD44        Fibroblasts   \n",
       "8329   CXCL12    CXCR4         CXCL12            CXCR4        Fibroblasts   \n",
       "8291     APOE     SDC2           APOE             SDC2        Fibroblasts   \n",
       "8443     APOE    ABCA1           APOE            ABCA1        Fibroblasts   \n",
       "8382     APOE    SORL1           APOE            SORL1        Fibroblasts   \n",
       "8368  ANGPTL2    ITGB1        ANGPTL2      ITGA5_ITGB1        Fibroblasts   \n",
       "8327   CXCL12    ITGAV         CXCL12            ITGAV        Fibroblasts   \n",
       "8447    TIMP2     CD44          TIMP2             CD44        Fibroblasts   \n",
       "8369  ANGPTL2     TLR4        ANGPTL2             TLR4        Fibroblasts   \n",
       "\n",
       "      ligand_stat  ligand_pvalue    ligand_padj  ligand_expr  ligand_props  \\\n",
       "4104   -34.265628  2.552207e-257  1.283528e-254     3.681040      0.999127   \n",
       "8424   -32.616964  2.357450e-233  9.790353e-231     3.606308      0.998240   \n",
       "8329    30.101009  4.700240e-199  1.229026e-196     2.238536      0.780211   \n",
       "8291    33.165903  3.340814e-241  1.814319e-238     2.378408      0.738561   \n",
       "8443    33.165903  3.340814e-241  1.814319e-238     2.378408      0.738561   \n",
       "8382    33.165903  3.340814e-241  1.814319e-238     2.378408      0.738561   \n",
       "8368   -32.630403  1.520047e-233  6.707206e-231     0.715059      0.482988   \n",
       "8327    30.101009  4.700240e-199  1.229026e-196     2.238536      0.780211   \n",
       "8447   -26.735568  1.817332e-157  3.054849e-155     1.093515      0.704145   \n",
       "8369   -32.630403  1.520047e-233  6.707206e-231     0.715059      0.482988   \n",
       "\n",
       "      ... receptor_pvalue  receptor_padj  receptor_expr  receptor_props  \\\n",
       "4104  ...    2.489514e-09   2.646970e-08       1.337293        0.762220   \n",
       "8424  ...    2.489514e-09   2.646970e-08       1.337293        0.762220   \n",
       "8329  ...    3.563692e-11   4.462986e-10       0.610545        0.376968   \n",
       "8291  ...    3.169624e-03   1.353437e-02       0.227950        0.146645   \n",
       "8443  ...    2.307151e-01   4.837241e-01       0.494338        0.356669   \n",
       "8382  ...    5.311427e-01   7.890612e-01       0.288026        0.225352   \n",
       "8368  ...    4.684623e-01   7.411197e-01       0.629548        0.463546   \n",
       "8327  ...    6.259711e-03   2.474708e-02       0.245828        0.198840   \n",
       "8447  ...    2.489514e-09   2.646970e-08       1.337293        0.762220   \n",
       "8369  ...    9.549634e-01   9.848294e-01       0.195188        0.163629   \n",
       "\n",
       "      interaction_stat  interaction_pvalue  interaction_padj  \\\n",
       "4104        -20.113885        1.244757e-09      1.323485e-08   \n",
       "8424        -19.289554        1.244757e-09      1.323485e-08   \n",
       "8329         18.361089        1.781846e-11      2.231493e-10   \n",
       "8291         18.058347        1.584812e-03      6.767183e-03   \n",
       "8443         17.182211        1.153576e-01      2.418620e-01   \n",
       "8382         16.896083        2.655714e-01      3.945306e-01   \n",
       "8368        -16.677693        2.342312e-01      3.705599e-01   \n",
       "8327         16.417433        3.129855e-03      1.237354e-02   \n",
       "8447        -16.348856        1.244757e-09      1.323485e-08   \n",
       "8369        -16.286964        4.774817e-01      4.924147e-01   \n",
       "\n",
       "      interaction_expr  interaction_props          interaction  \n",
       "4104          2.509166           0.880674             VIM^CD44  \n",
       "8424          2.471801           0.880230             VIM^CD44  \n",
       "8329          1.424540           0.578589         CXCL12^CXCR4  \n",
       "8291          1.303179           0.442603            APOE^SDC2  \n",
       "8443          1.436373           0.547615           APOE^ABCA1  \n",
       "8382          1.333217           0.481956           APOE^SORL1  \n",
       "8368          0.672304           0.473267  ANGPTL2^ITGA5_ITGB1  \n",
       "8327          1.242182           0.489526         CXCL12^ITGAV  \n",
       "8447          1.215404           0.733183           TIMP2^CD44  \n",
       "8369          0.455124           0.323308         ANGPTL2^TLR4  \n",
       "\n",
       "[10 rows x 22 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lr_res = lr_res.sort_values(\"interaction_stat\", ascending=False, key=abs)\n",
    "# Select all of the interactions with macrophages are the target\n",
    "lr_res = lr_res[lr_res[\"target\"] == \"Macrophages\"]\n",
    "lr_res.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "47be1f12-ad0d-43ff-b3c1-d4185b403a9d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "lr_res.to_csv(\"differential_expression_ligand_receptor_naive_v_resistant_target_macs.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "156f507d-ea9a-479e-b9d1-b4991d2604bc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "target\n",
       "Macrophages    1826\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lr_res[\"target\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01974bc3-e283-4b34-86f8-b721b635a914",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
